src/anndata/_core/merge.py

```diff
@@ -1567,10 +1679,17 @@ def concat(  # noqa: PLR0912, PLR0913, PLR0915
         )
         concat_annot.index = concat_indices
     else:
-        concat_annot = concat_dataset2d_on_annot_axis(annotations, join)
+        concat_annot = concat_dataset2d_on_annot_axis(
+            annotations, join, force_lazy=force_lazy
+        )
         concat_indices.name = DS_CONCAT_DUMMY_INDEX_NAME
+        concat_annot.index = concat_indices
     if label is not None:
-        concat_annot[label] = label_col
+        concat_annot[label] = (
+            label_col
+            if not isinstance(concat_annot, Dataset2D)
+            else (DS_CONCAT_DUMMY_INDEX_NAME, label_col)
+        )
 
```

Comment:
Can we move the previous implementation of `concat_annot[label] = label_col` into the `if are_any_annotations_dataframes` branch, and then move the xarray-specific logic into `concat_dataset2d_on_annot_axis`?

Separately, we should add a test for this, but if you want to leave that out of this PR, I can follow up anyway with something that tests both `label` and `keys` for xarray

Reply:
I think that is already tested in `test_concatenate.py::test_batch_key`, unless you mean something else?